File: pipeline/models/weapons.py
"""Pydantic model for weapons.json."""

from __future__ import annotations

from .common import LenientModel, StrictModel


class RangedWeapon(LenientModel):
    """A ranged weapon entry."""

    id: str
    name: str
    category: str
    type: str  # "pistol", "basic", "heavy", "thrown", "launcher"
    damage: str  # XkY notation
    damageType: str | None  # None for grenade/missile launchers (varies by ammo)
    pen: int
    rof: str  # e.g. "S/6", "S/-", "S/3/10"
    range: int | str  # Usually int; "Sx10" for thrown/compound weapons
    clip: int
    reload: str  # "Full", "2Full", "Free", etc.
    availability: str
    special: list[str]
    proficiency: list[str]
    description: str
    # Optional fields present on some entries
    notes: str | None = None


class MeleeWeapon(LenientModel):
    """A melee weapon entry."""

    id: str
    name: str
    category: str
    type: str  # "melee"
    damage: str  # XkY notation
    damageType: str
    pen: int
    availability: str
    special: list[str]
    proficiency: list[str]
    description: str
    # Optional fields present on some entries
    notes: str | None = None


class ThrownWeapon(LenientModel):
    """A thrown weapon entry (bolas, grenades, etc.)."""

    id: str
    name: str
    category: str
    type: str  # "thrown"
    damage: str  # XkY notation or "-" for non-damaging
    damageType: str | None  # None for smoke grenades, bolas
    pen: int
    range: int | str  # int (20) or formula string ("Sx3")
    availability: str
    special: list[str]
    proficiency: list[str]
    description: str


class WeaponsData(StrictModel):
    """Inner structure grouping weapons by type."""

    ranged: list[RangedWeapon]
    melee: list[MeleeWeapon]
    thrown: list[ThrownWeapon]


class WeaponsFile(StrictModel):
    """Top-level structure of weapons.json."""

    weapons: WeaponsData
    damageTypes: dict[str, str]
    qualities: dict[str, str]

File: src/lib/dtd/schemas/weapons.ts
/**
 * Zod schema for weapons data — mirrors pipeline/models/weapons.py.
 */

import { z } from "zod";

// ---------------------------------------------------------------------------
// RangedWeapon (lenient)
// ---------------------------------------------------------------------------

export const RangedWeapon = z
	.object({
		id: z.string(),
		name: z.string(),
		category: z.string(),
		type: z.string(),
		damage: z.string(),
		damageType: z.string().nullable(),
		pen: z.number().int(),
		rof: z.string(),
		range: z.union([z.number().int(), z.string()]),
		clip: z.number().int(),
		reload: z.string(),
		availability: z.string(),
		special: z.array(z.string()),
		proficiency: z.array(z.string()),
		description: z.string(),
		notes: z.string().nullish(),
	})
	.passthrough();
export type RangedWeapon = z.infer<typeof RangedWeapon>;

// ---------------------------------------------------------------------------
// MeleeWeapon (lenient)
// ---------------------------------------------------------------------------

export const MeleeWeapon = z
	.object({
		id: z.string(),
		name: z.string(),
		category: z.string(),
		type: z.string(),
		damage: z.string(),
		damageType: z.string(),
		pen: z.number().int(),
		availability: z.string(),
		special: z.array(z.string()),
		proficiency: z.array(z.string()),
		description: z.string(),
		notes: z.string().nullish(),
	})
	.passthrough();
export type MeleeWeapon = z.infer<typeof MeleeWeapon>;

// ---------------------------------------------------------------------------
// ThrownWeapon (lenient)
// ---------------------------------------------------------------------------

export const ThrownWeapon = z
	.object({
		id: z.string(),
		name: z.string(),
		category: z.string(),
		type: z.string(),
		damage: z.string(),
		damageType: z.string().nullable(),
		pen: z.number().int(),
		range: z.union([z.number().int(), z.string()]),
		availability: z.string(),
		special: z.array(z.string()),
		proficiency: z.array(z.string()),
		description: z.string(),
	})
	.passthrough();
export type ThrownWeapon = z.infer<typeof ThrownWeapon>;

// ---------------------------------------------------------------------------
// WeaponsData (strict)
// ---------------------------------------------------------------------------

export const WeaponsData = z
	.object({
		ranged: z.array(RangedWeapon),
		melee: z.array(MeleeWeapon),
		thrown: z.array(ThrownWeapon),
	})
	.strict();
export type WeaponsData = z.infer<typeof WeaponsData>;

// ---------------------------------------------------------------------------
// WeaponsFile (strict)
// ---------------------------------------------------------------------------

export const WeaponsFile = z
	.object({
		weapons: WeaponsData,
		damageTypes: z.record(z.string(), z.string()),
		qualities: z.record(z.string(), z.string()),
	})
	.strict();
export type WeaponsFile = z.infer<typeof WeaponsFile>;
